features/cellular/TARGET_GENERIC_MODEM/generic_modem_driver/ReferenceCellularInterface.h

```diff
@@ -0,0 +1,271 @@
+     *  the lookup table then the driver tries to resort to default APN settings.
+     *
+     *  Preferred method is to setup APN using 'set_credentials()' API.
+
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual nsapi_error_t connect();
+
+    /** Attempt to disconnect from the network
+     *
+     *  Brings down the network interface. Shuts down the PPP interface
+     *  of the underlying network stack. Does not bring down the Radio network
+     *
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual nsapi_error_t disconnect();
+
+    /** Adds or removes a SIM facility lock
+     *
+     * Can be used to enable or disable SIM pin check at device startup.
+     * This API sets up flags for the driver which would either enable or disable
+     * SIM pin checking depending upon the user provided argument while establishing
+     * connection. It doesn't do anything immediately other than setting up flags.
+     *
+     * @param check      can be set to true if the SIM pin check is supposed to be enabled
+     *                   and vice versa.
+     */
+    void check_sim_pin(bool check);
+
+    /** Change the pin for the SIM card
+     *
+     * Provide the new pin for your SIM card with this API. Old pin code will be assumed to
+     * be set using set_SIM_pin() API. This API have no immediate effect. While establishing
+     * connection, driver will change the SIM pin for the next boot.
+     *
+     * @param new_pin     new pin to be used in string format
+     */
+    void change_sim_pin(const char *new_pin);
+
+    /** Check if the connection is currently established or not
+     *
+     * @return true/false   If the cellular module have successfully acquired a carrier and is
+     *                      connected to an external packet data network using PPP, isConnected()
+     *                      API returns true and false otherwise.
+     */
+    virtual bool is_connected();
+
+    /** Get the local IP address
+     *
+     *  @return         Null-terminated representation of the local IP address
+     *                  or null if no IP address has been recieved
+     */
+    virtual const char *get_ip_address();
+
+    /** Get the local network mask
+     *
+     *  @return         Null-terminated representation of the local network mask
+     *                  or null if no network mask has been recieved
+     */
+    virtual const char *get_netmask();
+
+    /** Get the local gateways
+     *
+     *  @return         Null-terminated representation of the local gateway
+     *                  or null if no network mask has been recieved
+     */
+    virtual const char *get_gateway();
+
+    /** Get notified if the connection gets lost
+     *
+     *  @fptr         user defined callback
+     */
+    void connection_status_cb(void (*fptr)(nsapi_error_t));
```

Comment:
1. I don't think we should introduce this api now. There's a lot of work/discussion that is need around async apis.
2. If we do keep this, functions taking a callback should take the `Callback` class.